topics/bloomfilters.md

```diff
@@ -0,0 +1,128 @@
+
+### Ad placement
+
+Bloom filters can help answer the following questions to advertisers:
+* Has the user already seen this ad?
+* Has the user already bought this product?
+
+Use a Bloom filter for every user, storing all bought products. The recommendation engine can then suggest a new product and checks if the product is in the user's Bloom filter.
+
+* If no, the ad is shown to the user and is added to the Bloom filter.
+* If yes, the process restarts and repeats until it finds a product that is not present in the filter.
+
+### Check if URL's are malicious
+
+Bloom filters can answer the question "is a URL malicious?". Any URL inputted would be checked against a malicious URL bloom filter. 
+
+* If no then we allow access to the site
+* If yes then we can deny access or perform a full check of the URL
+
+### Check if a username is taken
+
+Bloom filters can answer the question: Has this username/email/domain name/slug already been used?
+
+For example for usernames. Use a Bloom filter for every username that has signed up. A new user types in the desired username. The app checks if the username exists in the Bloom filter.
+
+* If no, the user is created and the username is added to the Bloom filter.
+* If yes, the app can decide to either check the main database or reject the username.
+
+## Scaling and non scaling bloom filters
+
+The difference between scaling and non scaling bloom filters is that scaling bloom filters do not have a fixed capacity, but a capacity that can grow. While non-scaling bloom filters will have a fixed capacity which also means a fixed size. 
+
+When a scaling filter reaches its capacity, adding a new unique item will cause a new bloom filter to be created and added to the vector of bloom filters. This new bloom filter will have a larger capacity (previous bloom filter's capacity * expansion rate of the bloom object).
+
+When a non scaling filter reaches its capcity, if a user tries to add a new unique item an error will be returned
+
+The expansion rate is the rate that a scaling bloom filter will have its capacity increased by on the scale out. For example we have a bloom filter with capacity 100 at creation with an expansion rate of 2. After adding 101 unique items we will scale out and create a new filter with capacity 200. Then after adding 200 more unique items (301 items total) we will create a new filter of capacity 400 and so on. 
+
+### When should you use scaling vs non-scaling filters
+
+If the data size is known and fixed then using a non-scaling bloom filter is preferred, for example a static dictionary could use a non scaling bloom filter as the amount of items should be fixed. Likewise the reverse case for dynamic data and unknown final sizes is when you should use a scaling bloom filters.   
+
+## Default bloom properties
```

Comment:
Optional : We could add a monitoring section and briefly go over the INFO BF command response